server_addon/applications/client/ayon_applications/utils.py

```diff
@@ -281,13 +281,20 @@ def prepare_app_environments(
         app.environment
     ]
 
+    task_entity = data.get("task_entity")
     folder_entity = data.get("folder_entity")
     # Add tools environments
     groups_by_name = {}
     tool_by_group_name = collections.defaultdict(dict)
-    if folder_entity:
-        # Make sure each tool group can be added only once
-        for key in folder_entity["attrib"].get("tools") or []:
+    tools = None
+    if task_entity:
+        tools = task_entity["attrib"].get("tools")
+
+    if tools is None and folder_entity:
+        tools = folder_entity["attrib"].get("tools")
```

Comment:
Could someone explain to me when `tools` would be `None` on the task entity? Shouldn't it either always inherit from the parent folder to begin with OR have explicit overrides? As such, the task attribute for `tools` would always be the actual needed value and there would be no reason to check the folder entity if tools is None?

As such this would more likely be:
```suggestion
    if task_entity:
       tools = task_entity["attrib"].get("tools")
    else:
       tools = folder_entity["attrib"].get("tools")
```

Reply:
Folder entity might be optional too.
```python
tools = None
if task_entity:
    tools = task_entity["attrib"].get("tools")
elif folder_entity:
    tools = folder_entity["attrib"].get("tools")
```

Reply:
In that case the current merged code is already containing a bug since it doesn't use folder entity as an optional value.

Reply:
I think that's way more readable. 👍

Reply:
Should I make the commit again?